insert or update on table external sources violates foreign key constraint

juanignaciosl opened the pull request:
This fixes #12488.

mayoral commented on app/models/visualization/external_source.rb:
```diff
@@ -22,16 +22,7 @@ def initialize(visualization_id, import_url, geometry_types, rows_counted, size,
-      end
-
-      def importable_by(user)
+      def importable_by?(user)
```
😍

mayoral commented on app/services/visualization/common_data_service.rb:
```diff
@@ -78,11 +78,28 @@ def load_common_data_for_user(user, visualizations_api_url)
+                username: 'common-data'
+              )
+              # ActiveRecord array issue
+              external_source.update_attribute(:geometry_types, dataset['geometry_types'])
```
What error did you get from ActiveRecord if you try to set the field on the create method? Just curiosity.

juanignaciosl replied:
Generated SQL is wrong. It doesn't add single quotes around arrays. We searched for workarounds, but there's no one perfect for the gems versions that we use. This will be forgotten when we upgrade to Rails 4.

mayoral commented on spec/services/visualization/common_data_service_spec.rb:
```diff
@@ -50,6 +50,14 @@ def dataset(name, description: 'A very creative description')
     expect(remote_visualizations(@user).first.name).to eq 'ds1'
   end
 
+  it 'should import common data datasets within an ActiveRecord transaction (see #12488)' do
```
What's the expectation of this test? Not to raise an exception?

juanignaciosl replied:
Yes, that's what I explain in the next line: "# This would trigger an exception because of data integrity". It fails because of a FK (the one linked at the issue, indeed).

mayoral replied:
In that case we can do something like:

``` ruby
expect { service.load_common_data_for_user(@user, 'some_url') }.not_to raise_error CartoDB::Datasources::AuthError
```

juanignaciosl replied:
👍

mayoral commented on spec/services/visualization/common_data_service_spec.rb:
```diff
@@ -50,6 +50,16 @@ def dataset(name, description: 'A very creative description')
+        service.stubs(:get_datasets).returns([dataset('ds1')])
+        service.load_common_data_for_user(@user, 'some_url')
+      end
+    }.not_to raise_error
```
Nice! I usually prefer to tell here what error should not be raised instead of use the catch-all, but it works here.

juanignaciosl replied:
Yep, as the problem is quite indirect I prefer a broad spectrum check.

juanignaciosl commented:
Ey, I'm moving this backwards because after a last look I think that there's a redundant `.save!`.

juanignaciosl commented:
@mayoral sorry, I realized about a redundant `.save!`, and then I also realized that I could wrap the iteration through the data sources in a transaction, making it much faster. In addition, I've added a workaround that makes `change?` detection better, so we don't trigger unnecessary `UPDATE` requests. Could you do a second CR + Acceptance, please?

This involves user creation (normal signup, organization signup, organization creation, superadmin creation), and data library refresh (you can force it setting `last_common_data_update_date` to null and then opening the dashboard again).

Take a look at the queries in first insert, no metadata change, and changed metadata, I think they're close to optimal now. Let me know if you have trouble reproducing.

mayoral commented on app/models/carto/external_source.rb:
```diff
@@ -22,4 +23,14 @@ def update_data(import_url, geometry_types, rows_counted, size, username = nil)
+    array = [] if array.nil?
+    string = '{}' if string.nil?
+
+    "{#{array.join(',')}}" == string
```
Looks good, but I think this method (and maybe others) have more sense to be part of a generic Carto ActiveRecord base class that our models can use as a base class instead of `ActiveRecord::Base`.

I know, it's out of the scope of this issue. I'm just mentioning it to open the debate and consider to do it in the future.

juanignaciosl replied:
I wouldn't encourage inheritance for code reuse. Maybe a mixin. Anyway, agreed on refactor when we need this again. Maybe we'll upgrade versions before that and we can throw away this ugly hack.

mayoral replied:
👍
